### DISCUSSION TITLE
Simplify run-once root access setup script?

### DISCUSSION THREAD

**Discussion by seud0nym** [hack-technicolor/hack-technicolor#249] [Ideas]
The run-once root access setup script in https://hack-technicolor.readthedocs.io/en/stable/Upgrade/#preserving-root-access modifies `/etc/rc.local` and then deletes it once root access has been re-acquired. It also uses a complicated method to find firewall rules blocking port 22.

Is there a reason that a script in `/etc/uci-defaults` wouldn't be a better choice? It automatically does `uci commit`, runs before services are started, and auto-deletes on success.

Something like this would seem slightly simpler:

```bash
mkdir -p $(cat /proc/banktable/booted)/etc/uci-defaults
chmod 755 $(cat /proc/banktable/booted) $(cat /proc/banktable/booted)/etc $(cat /proc/banktable/booted)/etc/uci-defaults
cat <<-"END" > $(cat /proc/banktable/booted)/etc/uci-defaults/zzz-9999-reroot
  exec 1>/tmp/zzz-9999-reroot.out 2>/tmp/zzz-9999-reroot.err
  echo root:root | chpasswd
  sed -i 's#/root:.*$#/root:/bin/ash#' /etc/passwd
  sed -i -e 's/#//' -e 's#askconsole:.*$#askconsole:/bin/ash#' /etc/inittab
  for r in $(uci show firewall | grep "dest_port='22'" | cut -d. -f1-2); do 
    [ "$(uci get $r.target)" = DROP -a "$(uci -q get $r.src)" = lan -a -z "$(uci -q get $r.dest)" ] && uci set $r.target='ACCEPT'
  done
  uci add dropbear dropbear
  uci rename dropbear.@dropbear[-1]=afg
  uci set dropbear.afg.enable='1'
  uci set dropbear.afg.Interface='lan'
  uci set dropbear.afg.Port='22'
  uci set dropbear.afg.IdleTimeout='600'
  uci set dropbear.afg.PasswordAuth='on'
  uci set dropbear.afg.RootPasswordAuth='on'
  uci set dropbear.afg.RootLogin='1'
  uci set dropbear.lan.enable='0'
  /etc/init.d/dropbear enable
  exit 0
END
chmod +x $(cat /proc/banktable/booted)/etc/uci-defaults/zzz-9999-reroot
sync
``` 

I have successfully used a variation of this (my variation has some replaceable variables so it doesn't have a quoted `END` delimiter and has to escape the `$` signs) on version 17, 18, and 20 firmwares.

**Comment by LuKePicci** [comment]
It is not a bad idea to use UCI-defaults but the original intention was to keep that command's kit the same as we embed it into root tools/guides for all supported strategies. In these cases you have to run the UCI commit and you need to do the restart explicitly.

The firewall find & accept oneliner is needed for AFG where each command runs on a new shell and you have a char limit, that's why it is so horrible to read. Try to put your variation into AFG and see what happens.

**Comment by seud0nym** [comment]
Okay. Thanks for the explanation.
